torchmultimodal/models/albef.py

```diff
@@ -0,0 +1,244 @@
+        image_feat = F.normalize(self.vision_proj(image_embeds[:, 0, :]), dim=-1)
+        text_feat = F.normalize(self.text_proj(text_embeds[:, 0, :]), dim=-1)
+        return image_embeds, text_embeds, image_feat, text_feat
+
+    @torch.no_grad()
+    def _momentum_embeddings(self, image, text, text_atts):
+        self._momentum_update()
+        image_embeds_m = self.vision_encoder_m(image)
+        text_embeds_m = self.text_encoder_m(text, attention_mask=text_atts)
+        image_feat_m = F.normalize(self.vision_proj_m(image_embeds_m[:, 0, :]), dim=-1)
+        text_feat_m = F.normalize(self.text_proj_m(text_embeds_m[:, 0, :]), dim=-1)
+        return image_embeds_m, image_feat_m, text_feat_m
+
+    @torch.no_grad()
+    def _momentum_update(self):
+        for model, model_m in [
+            [self.vision_encoder, self.vision_encoder_m],
+            [self.text_encoder, self.text_encoder_m],
+            [self.multimodal_encoder, self.multimodal_encoder_m],
+            [self.vision_proj, self.vision_proj_m],
+            [self.text_proj, self.text_proj_m],
+        ]:
+            for param, param_m in zip(model.parameters(), model_m.parameters()):
+                param_m.data = param_m.data * self.momentum + param.data * (
+                    1 - self.momentum
+                )
+
+    @torch.no_grad()
+    def _update_queue(self, image_feat_m, text_feat_m):
+        image_feats, text_feats, _ = _gather_embeddings_and_labels(
+            image_feat_m, text_feat_m
+        )
+        batch_size = image_feats.shape[0]
+        ptr = int(self.queue_ptr)
+        assert self.queue_size % batch_size == 0
+
+        self.image_queue[:, ptr : ptr + batch_size] = image_feats.T
+        self.text_queue[:, ptr : ptr + batch_size] = text_feats.T
+        ptr = (ptr + batch_size) % self.queue_size
+        self.queue_ptr[0] = ptr
+
+    def _similarity(self, image_feat, text_feat, image_feat_m, text_feat_m, alpha):
+        with torch.no_grad():
+            image_feat_all = torch.cat(
+                [image_feat_m.t(), self.image_queue.clone().detach()], dim=1
+            )
+            text_feat_all = torch.cat(
+                [text_feat_m.t(), self.text_queue.clone().detach()], dim=1
+            )
+            sim_i2t_m = image_feat_m @ text_feat_all / self.temp
+            sim_t2i_m = text_feat_m @ image_feat_all / self.temp
+
+            sim_targets = torch.zeros(sim_i2t_m.size()).to(image_feat.device)
+            sim_targets.fill_diagonal_(1)
+
+            sim_i2t_targets = (
+                alpha * F.softmax(sim_i2t_m, dim=1) + (1 - alpha) * sim_targets
+            )
+            sim_t2i_targets = (
+                alpha * F.softmax(sim_t2i_m, dim=1) + (1 - alpha) * sim_targets
+            )
+
+        sim_i2t = image_feat @ text_feat_all / self.temp
+        sim_t2i = text_feat @ image_feat_all / self.temp
+        self._update_queue(image_feat_m, text_feat_m)
```

Comment:
question just for my understanding: why are the similarities not symmetric?

Reply:
It might be better to move the call to `update_queue` into `_momentum_update` instead for clarity (and I don't think either `image_feat_m` or `text_feat_m` gets modified between there and here).

Re @RdoubleA 's question, the similarity is computed by comparing the current batch's embeddings to those from the queue, so we want to calculate both `current_batch_image_embeds * full_queue_text_embeds^T` for the current batch's image similarity with all text embeddings and `current_batch_text_embeds * full_queue_image_embeds^T` for the current batch's text similarity with all image embeddings. Btw this is similar to what CLIP does in the distributed case ([ref](https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/modules/losses/contrastive_loss_with_temperature.py#L110-L117))